File: swift/Tests/SimpleTest.swift
@testable import Carbon

import XCTest

class SimpleTest: XCTestCase {
    var carbon: CarbonClient!
    var carbonWithToken: CarbonClient!

    override func setUpWithError() throws {
        self.carbon = CarbonClient(
            accessToken: nil,
            apiKey: "API_KEY",
            customerId: "CUSTOMER_ID",
            basePath: "http://127.0.0.1:4010"
        )
    }

    override func tearDownWithError() throws {}

    func testAuthMethods() async throws {
        let expectation = XCTestExpectation(description: "Token is not null")
        var token: TokenResponse?

        do {
            let tokenResponse = try await self.carbon.auth.getAccessToken()
            token = tokenResponse
            expectation.fulfill()
        } catch {
            print(error)
        }
        
        await fulfillment(of: [expectation], timeout: 10.0)

        self.carbonWithToken = CarbonClient(
            accessToken: token!.accessToken,
            apiKey: nil,
            customerId: nil,
            basePath: "http://127.0.0.1:4010"
        )

        let expectation2 = XCTestExpectation(description: "White labeling is not null")
        do {
            let _ = try await self.carbonWithToken.auth.getWhiteLabeling()
            expectation2.fulfill()
        } catch {
            print(error)
        }
        await fulfillment(of: [expectation2], timeout: 10.0)
    }

    func testVariousEndpoints() async throws {
        let expectation = XCTestExpectation(description: "API call successful")
        do {
            let pagination = Pagination(limit: 10, offset: 0)
            let orderBy = OrganizationUserDataSourceOrderByColumns.createdAt
            let orderDir = OrderDir.desc
            let filter = OrganizationUserDataSourceFilters(source: .googleDrive)
            let result = try await self.carbon.dataSources.queryUserDataSources(pagination: pagination, orderBy: orderBy, orderDir: orderDir, filters: filter)
            print(result)
            expectation.fulfill()
        } catch {
            print(error)
        }
        await fulfillment(of: [expectation], timeout: 10.0)

        let expectation2 = XCTestExpectation(description: "API call successful")
        do {
            let fileInput = S3GetFileInput(id: "ID", bucket: "BUCKET")
            let result = try await self.carbon.integrations.syncS3Files(ids: [fileInput])
            print(result)
            expectation2.fulfill()
        } catch {
            print(error)
        }
        await fulfillment(of: [expectation2], timeout: 10.0)

        let expectation3 = XCTestExpectation(description: "API call successful")
        do {
            let url = URL(fileURLWithPath: "./README.md")
            let result = try await self.carbon.files.upload(file: url)
            print(result)
            expectation3.fulfill()
        } catch {
            print(error)
        }
        await fulfillment(of: [expectation3], timeout: 10.0)

        let expectation4 = XCTestExpectation(description: "API call successful")
        do {
            let filters = OrganizationUserFilesToSyncFilters(ids: [1,2,3])
            let result = try await self.carbon.files.deleteV2(filters: filters, sendWebhook: false)
            print(result)
            expectation4.fulfill()
        } catch {
            print(error)
        }
        await fulfillment(of: [expectation4], timeout: 10.0)
    }
}

File: java/src/test/java/com/konfigthis/carbonai/client/SimpleTest.java
package com.konfigthis.carbonai.client;

import static org.junit.Assert.assertNotNull;
import com.konfigthis.carbonai.client.model.TokenResponse;
import com.konfigthis.carbonai.client.model.WhiteLabelingResponse;
import com.konfigthis.carbonai.client.model.OrganizationUserDataSourceResponse;
import com.konfigthis.carbonai.client.model.Pagination;
import com.konfigthis.carbonai.client.model.OrganizationUserDataSourceFilters;
import com.konfigthis.carbonai.client.model.DataSourceTypeNullable;
import com.konfigthis.carbonai.client.model.OrganizationUserDataSourceOrderByColumns;
import com.konfigthis.carbonai.client.model.OrderDir;
import com.konfigthis.carbonai.client.model.GenericSuccessResponse;
import com.konfigthis.carbonai.client.model.S3GetFileInput;
import com.konfigthis.carbonai.client.model.UserFile;

import java.util.ArrayList;
import java.util.List;
import java.io.File;
import org.junit.jupiter.api.Test;

public class SimpleTest {
    final String MOCK_SERVER_URL = "http://localhost:4010";

    @Test
    public void initClientTest() {
        Configuration configuration = new Configuration();
        configuration.host = MOCK_SERVER_URL;
        configuration.apiKey  = "YOUR API KEY";
        configuration.customerId  = "YOUR API KEY";
        Carbon client = new Carbon(configuration);
        assertNotNull(client);
    }

    @Test
    public void authTest() {
        Configuration configuration = new Configuration();
        configuration.host = MOCK_SERVER_URL;
        configuration.apiKey = "API KEY";
        configuration.customerId = "CUSTOMER ID";
        Carbon client = new Carbon(configuration);

        TokenResponse tr = null;
        try {
            tr = client.auth.getAccessToken().execute();
        } catch (ApiException e) {}
        assertNotNull(tr);
        String token = tr.getAccessToken();
        assertNotNull(token);

        Configuration configuration2 = new Configuration();
        configuration2.host = MOCK_SERVER_URL;
        configuration2.accessToken = token;
        Carbon client2 = new Carbon(configuration2);
        WhiteLabelingResponse wlr = null;
        try {
            wlr = client2.auth.getWhiteLabeling().execute();
        } catch (ApiException e) {}
        assertNotNull(wlr);
    }

    @Test
    public void endpointTest() {
        Configuration configuration = new Configuration();
        configuration.host = MOCK_SERVER_URL;
        configuration.apiKey = "API KEY";
        configuration.customerId = "CUSTOMER ID";
        Carbon client = new Carbon(configuration);

        OrganizationUserDataSourceResponse oudsr = null;
        try {
            oudsr = client.dataSources.queryUserDataSources()
                                       .pagination(new Pagination().limit(10).offset(0))
                                       .orderBy(OrganizationUserDataSourceOrderByColumns.CREATED_AT)
                                       .orderDir(OrderDir.DESC)
                                       .filters(new OrganizationUserDataSourceFilters().source(DataSourceTypeNullable.GOOGLE_DRIVE))
                                       .execute();
        } catch (ApiException e) {}
        assertNotNull(oudsr);

        GenericSuccessResponse gsr = null;
        List<S3GetFileInput> s3GetFileInput = new ArrayList<S3GetFileInput>();
        s3GetFileInput.add(new S3GetFileInput().id("ID").bucket("BUCKET"));
        try {
            gsr = client.integrations.syncS3Files(s3GetFileInput).execute();
        } catch (ApiException e) {}
        assertNotNull(gsr);

        UserFile uf = null;
        File file = new File("README.md");
        try {
            uf = client.files.upload(file).execute();
        } catch (ApiException e) {}
        assertNotNull(uf);
    }
}
